tensorflow/python/ops/nn_ops.py

```diff
@@ -2142,6 +2142,28 @@ def xw_plus_b_v1(x, weights, biases, name=None):  # pylint: disable=invalid-name
     mm = math_ops.matmul(x, weights)
     return bias_add_v1(mm, biases, name=name)
 
+def _get_noise_shape(x, noise_shape):
+  # If noise_shape is none return immediately.
+  if noise_shape is None:
+    return array_ops.shape(x)
+
+  try:
+    # Best effort to figure out the intended shape.
+    # If not possible, let the op to handle it.
+    noise_shape_ = tensor_shape.as_shape(noise_shape)
+    if (x.shape.dims is not None and
+        len(x.shape.dims) == len(noise_shape_.dims)):
+      new_dims = []
+      for i, dim in enumerate(x.shape.dims):
+        if noise_shape_.dims[i].value is None and dim.value is not None:
+          new_dims.append(dim.value)
+        else:
+          new_dims.append(noise_shape_.dims[i].value)
+      return tensor_shape.TensorShape(new_dims)
+  except (TypeError, ValueError):
```

Comment:
Under what condition does a TypeError or a ValueError show up?

Reply:
@vrv In case of eager mode TypeError and ValueError will show up:
```
`Tensor` objects are not iterable when eager execution is not enabled. To iterate over this tensor use `tf.map_fn`
```
This is triggered by 
```
noise_shape_ = tensor_shape.as_shape(noise_shape)
```
from the earlier step.

Reply:
Can you put the try except around just the as_shape() function to scope its protection more carefully?

Or maybe just disable the entire codepath if in_eager_mode?   cc @akshayka in case he knows under what conditions we should disable eager codepaths.

Reply:
@vrv Updated so that try except is around `as_shape()`.